librz/util/str_search.c

```diff
@@ -0,0 +1,301 @@
+#include <rz_util/rz_str_search.h>
+#include <rz_util/rz_utf8.h>
+#include <rz_util/rz_utf16.h>
+#include <rz_util/rz_utf32.h>
+
+typedef enum {
+	SKIP_STRING,
+	RETRY_ASCII,
+	STRING_OK,
+} FalsePositiveResult;
+
+RZ_API void rz_detected_string_free(RzDetectedString *str) {
+	if (str) {
+		free(str->string);
+		free(str);
+	}
+}
+
+static FalsePositiveResult reduce_false_positives(const RzUtilStrScanOptions *opt, ut8 *str, int size, RzStrEnc str_type) {
+	int i, num_blocks, *block_list;
+	int *freq_list = NULL, expected_ascii, actual_ascii, num_chars;
+
+	switch (str_type) {
+	case RZ_STRING_ENC_LATIN1: {
+		for (i = 0; i < size; i++) {
+			char ch = str[i];
+			if (ch != '\n' && ch != '\r' && ch != '\t') {
+				if (!IS_PRINTABLE(str[i])) {
+					return SKIP_STRING;
+				}
+			}
+		}
+		break;
+	}
+	case RZ_STRING_ENC_UTF8:
+	case RZ_STRING_ENC_UTF16LE:
+	case RZ_STRING_ENC_UTF32LE:
+		num_blocks = 0;
+		block_list = rz_utf_block_list((const ut8 *)str, size - 1,
+			str_type == RZ_STRING_ENC_UTF16LE ? &freq_list : NULL);
+		if (block_list) {
+			for (i = 0; block_list[i] != -1; i++) {
+				num_blocks++;
+			}
+		}
+		if (freq_list) {
+			num_chars = 0;
+			actual_ascii = 0;
+			for (i = 0; freq_list[i] != -1; i++) {
+				num_chars += freq_list[i];
+				if (!block_list[i]) { // ASCII
+					actual_ascii = freq_list[i];
+				}
+			}
+			free(freq_list);
+			expected_ascii = num_blocks ? num_chars / num_blocks : 0;
+			if (actual_ascii > expected_ascii) {
+				free(block_list);
+				return RETRY_ASCII;
+			}
+		}
+		free(block_list);
+		if (num_blocks > opt->max_uni_blocks) {
+			return SKIP_STRING;
+		}
+		break;
+	default:
+		return STRING_OK;
+	}
+
+	return STRING_OK;
+}
+
+static ut64 adjust_offset(RzStrEnc str_type, ut8 *buf, const ut64 str_start) {
+	switch (str_type) {
+	case RZ_STRING_ENC_UTF16LE:
+		if (str_start > 1) {
+			const ut8 *p = buf + str_start - 2;
+			if (p[0] == 0xff && p[1] == 0xfe) {
+				return 2; // \xff\xfe
+			}
+		}
+		break;
+	case RZ_STRING_ENC_UTF16BE:
+		if (str_start > 1) {
+			const ut8 *p = buf + str_start - 2;
+			if (p[0] == 0xfe && p[1] == 0xff) {
+				return 2; // \xfe\xff
+			}
+		}
+		break;
+	case RZ_STRING_ENC_UTF32LE:
+		if (str_start > 3) {
+			const ut8 *p = buf + str_start - 4;
+			if (p[0] == 0xff && p[1] == 0xfe && !p[2] && !p[3]) {
+				return 4; // \xff\xfe\x00\x00
+			}
+		}
+		break;
+	case RZ_STRING_ENC_UTF32BE:
+		if (str_start > 3) {
+			const ut8 *p = buf + str_start - 4;
+			if (!p[0] && !p[1] && p[2] == 0xfe && p[3] == 0xff) {
+				return 4; // \x00\x00\xfe\xff
+			}
+		}
+		break;
+	default:
+		break;
+	}
+
+	return 0;
+}
+
+RZ_API int rz_scan_strings(RzBuffer *buf_to_scan, RzList *list, const RzUtilStrScanOptions *opt,
```

Comment:
Please add Doxygen description of the function, it is especially important in regard to the parameters and `opt` description. Also the return values.

Reply:
I added the description to the header file, I have to add it here?

Reply:
Yes, please move it from the header to here.